# Bài tập 5

Nguyễn Trọng Khiêm - 1712529

## Câu 1 (1 điểm)

Xét đẳng thức đề cho:  
$$\mathbb{E}_{\mathcal{D}}[E_{in}(\mathbf{w}_{lin})] = \sigma^2\left(1 - \frac{d+1}{N}\right)$$  
Đặt $VT = A$, ta được:  
\begin{eqnarray*}
A & = & \sigma^2\left(1 - \frac{d+1}{N}\right)\\ 
 \frac{A}{\sigma^2} & = & 1 - \frac{d+1}{N}\\
\frac{d+1}{N} & = & 1 - \frac{A}{\sigma^2}\\
 N & = & \frac{d+1}{1 - \frac{A}{\sigma^2}}
\end{eqnarray*}
Mà đề bài cho $\sigma=0.1, d=8$, và hỏi cần $N$ nhỏ nhất là bao nhiêu để $E_{in} > 0.008$. Do đó, ta suy ra:  
$$ N > \frac{d+1}{1 - \frac{A}{\sigma^2}} = \frac{9}{1 - \frac{0.008}{0.1^2}}$$ 
$$\Rightarrow N > 45$$  
Do đó N cần tìm phải nhỏ nhất và đồng thời phải lớn hơn 45, nên đáp án sẽ là 100.

Vậy em chọn đáp án [c] 100.

## Câu 2 (1 điểm)

Ta có: $h(\mathbf{x}) = sign(\tilde{w}_0 + \tilde{w}_1 x_1^2 + \tilde{w}_2 x_2^2)$  
Khi $x_1$ siêu âm hoặc siêu dương thì $x_1^2$ sẽ dương và rất lớn, và  $\tilde{w}_1 x_1^2$ sẽ là đại lượng quyết định dấu của tổng trong công thức ở trên. Do $x_1^2$ dương nên $\tilde{w}_1$ sẽ quyết định dấu của $\tilde{w}_1 x_1^2$. Mà khi $x_1$ siêu dương, với $x_2$ có giá trị tuyệt đối nhỏ (không cần biết âm hay dương), $h(x)$ cho giá trị âm, do đó bắt buộc $\tilde{w}_1$ phải mang giá trị âm, hay  $\tilde{w}_1 < 0$.  
Tương tự, khi $x_2$ siêu âm hoặc siêu dương, $h(x)$ đều cho giá trị dương, nên bắt buộc $\tilde{w}_2$ phải mang giá trị dương, hay  $\tilde{w}_2 > 0$.

Vậy em chọn đáp án [d].

## Câu 3 (1 điểm)

Xét mô hình tuyến tính,ban đầu ta có $d_{VC} = d+1 = 2 + 1 =3$ (2D perceptron).  
Tuy nhiên, sau khi biến đổi sang không gian $\Phi_4$, $d_{VC} \leq \tilde{d} +1 = 14 + 1 =15$.  
Do đó giá trị nhỏ nhất không nhỏ hơn $d_{VC}$ là 15.

Vậy em chọn đáp án [c] 15.

## Câu 4 (1 điểm)

$$\frac{\partial E}{\partial u} = 2(e^v + 2ve^{-u}) (ue^v - 2ve^{-u})$$
$$\frac{\partial E}{\partial v} = 2(ue^v - 2e^{-u})(ue^v - 2ve^{-u})$$

Vậy em chọn đáp án [e].

## Câu 5 (1 điểm)

In [1]:
# Cau 5 + 6
import numpy as np


def E(x):
    """
    E(u, v) = (u*e^v - 2v*e^(-u))^2
    """
    u = x[0]
    v = x[1]
    return (u*np.exp(v) - 2*v*np.exp(-u))**2

def gradE(x):
    """
    Hàm tính đạo hàm riêng của E(u, v) theo u và v theo công thức ở câu 4.
    """
    u = x[0]
    v = x[1]
    return np.array([
        2*(u*np.exp(v) - 2*v*np.exp(-u))*(np.exp(v) + 2*v*np.exp(-u)),
        2*(u*np.exp(v) - 2*v*np.exp(-u))*(u*np.exp(v) - 2*np.exp(-u))
    ])

def gradient_descent(E, gradE, x0, threshold, l_rate):
    x = x0.copy()
    n_iters = 0
    err = E(x)
    us = [x[0]]
    vs = [x[1]]
    while err > threshold:
        grad = gradE(x)
        x = x - l_rate*grad
        us.append(x[0])
        vs.append(x[1])
        err = E(x)
        n_iters += 1
        if n_iters > 100:
            print("Không hội tụ!!!")
            return
    return (x, err, n_iters, us, vs)
    
    
x0 = np.array([1, 1])
threshold = 10e-14
l_rate = 0.1
final_uv, err, n_iters, us, vs = gradient_descent(E, gradE, x0, threshold, l_rate)
print("Final (u, v) =", final_uv)
print("Final err:", err)
print("Số vòng lặp:", n_iters)

Final (u, v) = [0.04473629 0.02395871]
Final err: 1.2086833944220747e-15
Số vòng lặp: 10


Vậy em chọn đáp án [d].

## Câu 6 (1 điểm)

Kết quả (u,v) = (0.04473629 0.02395871) gần với (0.045, 0.024).  
Vậy em chọn đáp án [e].

## Câu 7 (1 điểm)

In [2]:
def E(u, v):
    return (u*np.exp(v) - 2*v*np.exp(-u))**2

def dE_du(u, v):
    return 2*(u*np.exp(v) - 2*v*np.exp(-u))*(np.exp(v) + 2*v*np.exp(-u))

def dE_dv(u, v):
    return 2*(u*np.exp(v) - 2*v*np.exp(-u))*(u*np.exp(v) - 2*np.exp(-u))

def coordinate_descent(E, dE_du, dE_dv, u0, v0, threshold, l_rate):
    u, v = u0, v0
    us = [u]
    vs = [v]
    for it in range(15):
        # step1 move only along the u coordinate
        u = u - l_rate*dE_du(u, v)
        us.append(u)
        vs.append(v)
        # step2 move only along the v coordinate
        v = v - l_rate*dE_dv(u, v)
        us.append(u)
        vs.append(v)
    return u, v, E(u, v), us, vs


u0 = 1.0
v0 = 1.0
threshold = 10e-14
l_rate = 0.1
final_u, final_v, err, us, vs = coordinate_descent(E, dE_du, dE_dv, u0, v0, threshold, l_rate)
print("Final (u, v):", final_u, final_v)
print("Final err:", err)

Final (u, v): 6.29707589930517 -2.852306954077811
Final err: 0.13981379199615315


Kết quả nằm trong khoảng 0.10  
vậy em chọn đáp án [a].

## Câu 8 (1 điểm)

In [3]:
def random_point():
    return 2*np.random.rand()-1, 2*np.random.rand()-1

def random_line():
    x0, y0 = random_point()
    x1, y1 = random_point()
    a = (y1 - y0)/(x1 - x0)
    b = y1 - a*x1
    return b, a

def generate_dataset(N, b, a):
    X = 2*np.random.rand(N, 3)-1
    X[:,0] = 1
    y = np.sign(b + a*X[:,1] - X[:,2])
    return X, y

    return 1 / (1 + np.exp(-x))

class LogisticModel:
    
    def fit(self, X, y):
        eta = 0.01
        N = X.shape[0]
        w = np.zeros(X.shape[1])
        prev_w = np.ones(X.shape[1])
        epoch = 0
        
        while True:
            prev_w = w
            for n in np.random.permutation(N):
                n = np.random.randint(0, N)
                gradE = -y[n]*X[n,:] / (1 + np.exp(y[n] * w.dot(X[n,:])))
                w = w - eta * gradE

            epoch += 1
            if np.linalg.norm(w - prev_w) < 0.01:
                self.w = w
                break
                
        return epoch
            
    
    def predict(self, X):
        return np.sign(X @ self.w)
    
    def cross_entropy_error(self, X, y):
        N = X.shape[0]
        errs = []
        for n in range(N):
            errs.append(np.log(1 + np.exp(-y[n] * self.w.dot(X[n,:]))))
        return np.mean(errs)
    
num_epochs = np.array([])
cross_entropy_errors = np.array([])
        
for experiment in range(1, 100):
    b, a = random_line()
    X, y = generate_dataset(100, b, a)
    log_model = LogisticModel()
    epochs = log_model.fit(X, y)

    X_test, y_test = generate_dataset(2000, b, a)
    y_predicted = log_model.predict(X_test)

    err = log_model.cross_entropy_error(X_test, y_test)
    num_epochs = np.concatenate((num_epochs, [epochs])) 
    cross_entropy_errors = np.concatenate((cross_entropy_errors, [err]))
    
print("Average Epochs: ", np.average(num_epochs))
print("Average Error: ", np.average(cross_entropy_errors))

Average Epochs:  332.54545454545456
Average Error:  0.10036756190560842


Vậy em chọn đáp án [d].

## Câu 9 (1 điểm)

Khoảng 330 epochs.  
Vậy em chọn đáp án [a].

## Câu 10 (1 điểm)

- PLA: 
    - Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ được phân lớp đúng thì $\mathbf{w}$ không thay đổi.
    - Với mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ bị phân lớp sai thì cập nhật: $\mathbf{w} \leftarrow \mathbf{w} + y\mathbf{x}$.

- SGD: với mỗi mẫu $\left(\mathbf{x}^{(n)}, y^{(n)}\right)$ thì đều cập nhật $\mathbf{w} \leftarrow \mathbf{w} - \eta \nabla e(\mathbf{w})$.  
Để đơn giản, ta xét với $\eta=1$.  
Giả sử kết quả đúng là $e(\mathbf{w}) = -y\mathbf{w}^T\mathbf{x}$.  
Khi đó, gradient sẽ là: $\nabla e = -y\mathbf{x}$.  
Thay vào công thức cập nhật SGD, ta được: $$\mathbf{w} \leftarrow \mathbf{w} + y\mathbf{x}$$  
Kết quả đã giống với PLA. Tuy nhiên, nó chỉ đúng với những điểm phân lớp sai, vì PLA chỉ cập nhật như vậy với điểm phân lớp sai. Còn ở đây, SGD áp dụng cập nhật cho mọi điểm.  
Để có thể cập nhật đúng cho các điểm đã phân lớp đúng, ta sẽ chọn:  
$$e(\mathbf{w}) = - min(0, y \mathbf{w}^T \mathbf{x})$$  
vì với các điểm đã phân lớp đúng, giá trị của $y \mathbf{w}^T \mathbf{x}$ là dương, khi đó:
$$e(\mathbf{w}) = - min(0, y \mathbf{w}^T \mathbf{x}) = 0$$  
Điều này có nghĩa là với những điểm đã phân lớp đúng gradient $\nabla e = \nabla 0 = 0$, hay những điểm phân lớp đúng sẽ không đóng góp vào quá trình cập nhật $\mathbf{w}$.  
Vậy kết quả cần tìm là:
$$e(\mathbf{w}) = - min(0, y \mathbf{w}^T \mathbf{x})$$  

Vậy em chọn đáp án [e].